Mathlib/Tactic/ComputeDegree.lean

```diff
@@ -4,37 +4,461 @@ Released under Apache 2.0 license as described in the file LICENSE.
+let (can_progress, curr_static) := ← mvs.partitionM fun mv => do
+  return dispatchLemma (twoHeadsArgs (← mv.getType'')) != ``id
+let progress := ← can_progress.mapM fun mv => do
+  let lem := dispatchLemma <| twoHeadsArgs (← mv.getType'')
+  mv.applyConst <| lem
+return (progress.join, static ++ curr_static)
+
+/-- `miscomputedDegree? deg false_goals` takes as input
+*  an `Expr`ession `deg`, representing the degree of a polynomial
+   (i.e. an `Expr`ession of inferred type either `ℕ` or `WithBot ℕ`);
+*  a list of `MVarId`s `false_goals`.
+
+Although inconsequential for this function, the list of goals `false_goals` reduces to `False`
+if `norm_num`med.
+`miscomputedDegree?` extracts error information from goals of the form
+*  `a ≠ b`, assuming it comes from `⊢ coeff_of_given_degree ≠ 0`
+   -- reducing to `False` means that the coefficient that was supposed to vanish, does not;
+*  `a ≤ b`, assuming it comes from `⊢ degree_of_subterm ≤ degree_of_polynomial`
+   -- reducing to `False` means that there is a term of degree that is apparently too large;
+*  `a = b`, assuming it comes from `⊢ computed_degree ≤ given_degree`
+   -- reducing to `False` means that there is a term of degree that is apparently too large.
+
+The cases `a ≠ b` and `a = b` are not a perfect match with the top coefficient:
+reducing to `False` is not exactly correlated with a coefficient being non-zero.
+It does mean that `compute_degree` reduced the initial goal to an unprovable state
+(unless there was already a contradiction in the initial hypotheses!), but it is indicative that
+there may be some problem.
+-/
+def miscomputedDegree? (deg : Expr) : List Expr → List MessageData
+  | tgt::tgts =>
+    let rest := miscomputedDegree? deg tgts
+    if tgt.ne?.isSome then
+      m!"* the coefficient of degree {deg} may be zero" :: rest
+    else if let some ((Expr.const ``Nat []), lhs, _) := tgt.le? then
+      m!"* there is at least one term of naïve degree {lhs}" :: rest
+    else if let some (_, lhs, _) := tgt.eq? then
+      m!"* there may be a term of naïve degree {lhs}" :: rest
+    else rest
+  | [] => []
+
+/--
+`compute_degree` is a tactic to solve goals of the form
+*  `natDegree f = d`,
+*  `degree f = d`,
+*  `natDegree f ≤ d`,
+*  `degree f ≤ d`,
+*  `coeff f d = r`, if `d` is the degree of `f`.
+
+The tactic may leave goals of the form `d' = d` `d' ≤ d`, or `r ≠ 0`, where `d'` in `ℕ` or
+`WithBot ℕ` is the tactic's guess of the degree, and `r` is the coefficient's guess of the
+leading coefficient of `f`.
+
+`compute_degree` applies `norm_num` to the left-hand side of all side goals, trying to clos them.
+
+The variant `compute_degree!` first applies `compute_degree`.
+Then it uses `norm_num` on all the whole remaining goals and tries `assumption`.
+-/
+syntax (name := computeDegree) "compute_degree" "!"? "-debug"? : tactic
+
+@[inherit_doc computeDegree]
+macro "compute_degree!" "-debug"? : tactic => `(tactic| compute_degree ! -debug)
```

Comment:
Let's not use a debug flag. Instead let's us the standard interface of a trace option.
```suggestion
syntax (name := computeDegree) "compute_degree" "!"? : tactic

initialize registerTraceClass `Tactic.compute_degree
```